explanation/journey-of-a-change-to-production.rst

```diff
@@ -0,0 +1,155 @@
+  types of database changes are applied through hot patches. There are some differences
+  between these and let us focus on the cold patches for the time being.
+
+* **So what happens after the database schema changes submitted in a merge proposal, reviewed, and
+  approved?**
+
+  Once a database patch is approved and the status of the corresponding merge proposal is set
+  to "Approved", a `Jenkins trigger <https://jenkins.ols.ps5.canonical.com/job/trigger-launchpad/>`_
+  runs a `Jenkins job <https://jenkins.ols.ps5.canonical.com/job/launchpad/>`_ that checks for
+  unmerged merge proposals that are approved, and merges them to their target branch. So this
+  job will merge the approved database merge proposal. The Jenkins jobs are defined in the
+  `ols-jenkaas-jobs <https://git.launchpad.net/ols-jenkaas-jobs>`_ repository.
+
+* **Now my changes are merged, how can I deploy them to production?**
+
+  There are multiple steps/processes that need to happen before a database patch can be
+  deployed to production. Now that the database patch has been merged to the ``db-devel``
+  branch, the ``lp-db-devel-focal`` job on buildbot runs the full Launchpad test suite
+  on the ``db-devel`` branch. Once the tests pass, there is a cron job (TODO: add more details)
+  that runs on the buildbot master unit which checks for new commits that have completed
+  a successful buildbot run and merges them to the ```db-stable`` branch.
+
+  Once this is done, the new commits to the ``db-stable`` branch that have not yet been
+  deployed to production show up on the Deployable page, https://deployable.ols.canonical.com/projects/launchpad-db.
+
+  There is yet another Jenkins job on the OLS Jenkins instance that checks for new commits
+  to the ``db-stable`` branch and builds a new Launchpad payload for the latest revision in
+  the ``db-stable`` branch, and publishes to Swift, at a well-known location whose path contains
+  the hash of the commit that the tarball was built for.
+
+* **How can I test my database schema changes before attempting to deploy them to production?**
+
+  Launchpad as 2 pre-production environments, ``qastaging`` and ``staging``. The former is primarily
+  used for testing code changes whereas the latter is used for testing the database changes.
+
+  Launchpad is deployed using Juju charms using ``mojo``, a convenience wrapper around Juju that provides
+  a lot of useful functionality. The ``mojo`` configuration for deploying Launchpad to various environments
+  is defined in the ``lp:launchpad-mojo-specs`` repository, under the ``lp`` subdirectory. The ``bundle.yaml``
+  file in that directory is a Jinja2 template that gets rendered to an appropriate Juju bundle.yaml for each
+  environment. This ``bundle.yaml`` file contains 2 different ``build_label`` values - the hash of the ``stable``
+  branch commit that is deployed to production and also the hash of the ``db-stable`` branch commit that is
+  deployed to production.
```

Comment:
Sorry, it confused me a bit here whether the `build_label` is only used for deployments on production, or production and staging both.  

I think the earlier line states that `bundle.yaml` is generated according to environment but it would be better to not say the `build_label` values are only for production. 

Please correct me if I understood anything wrong.

Reply:
The `build_label` is always used in all deployments. In production, we use what is defined in the `launchpad-mojo-specs` repository. In staging and qastaging, we override it locally to the latest commits in the `db-stable` and `stable` branches respectively before invoking `mojo run`. See https://git.launchpad.net/launchpad-bastion-scripts/tree/auto-upgrade-staging#n32 and https://git.launchpad.net/launchpad-bastion-scripts/tree/auto-upgrade-qastaging#n32

Reply:
I have reworked this documentation from scratch and some of the minute details like this have been removed. Please re-review.